File: C++/20.观察者模式/HexaObserver.cpp
/**
 * @file HexaObserver.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-21
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "HexaObserver.h"

#include <bitset>
#include <iostream>

#include "Subject.h"

HexaObserver::HexaObserver(std::shared_ptr<Subject> subject) {
    subject_ = subject;
    subject_->Attach(this);
}

HexaObserver::~HexaObserver() {}

void HexaObserver::Update() { std::cout << "Hex String: " << std::hex << subject_->GetState() << std::endl; }

File: Objective-C/20.ObserverPattern/HexaObserver.m
//
//  HexaObserver.m
//  20.ObserverPattern
//
//  Created by Layne on 2020/11/13.
//

#import "HexaObserver.h"
#import "Subject.h"

@implementation HexaObserver {
    Subject* _subject;
}

- (instancetype)initWithSubject:(Subject*)subject {
    self = [super init];
    if (self) {
        _subject = subject;
        [_subject attach:self];
    }
    return self;
}

- (void)update {
    NSLog(@"Hex String %X", [_subject getState]);
}

@end
